File: jtablayout_lib/src/main/java/com/liang/jtab/Tab.java
package com.liang.jtab;

import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.annotation.DrawableRes;
import android.support.annotation.IdRes;
import android.view.View;

public interface Tab {

    /**
     * Setting Tab's Arrangement
     *
     * @param mode
     */
    Tab setOrientationMode(int mode);

    int getPosition();

    void setPosition(int position);

    CharSequence getTitle();

    /**
     * Set the title of Tab.
     *
     * @param title
     */
    Tab setTitle(CharSequence title);

    Object getContentDesc();

    /**
     * Tab's custom extension content
     *
     * @param contentDesc
     */
    Tab setContentDesc(Object contentDesc);

    ColorStateList getTitleColor();

    /**
     * Tab Title Switching Colors
     *
     * @param defaultColor
     * @param selectedColor
     */
    Tab setTitleColor(@ColorInt int defaultColor, @ColorInt int selectedColor);

    /**
     * Tab Title Switching Colors
     *
     * @param titleColor
     */
    Tab setTitleColor(ColorStateList titleColor);

    Tab setBackgroundRes(@DrawableRes int resId);

    Tab setBackgroundDraw(Drawable background);

    Drawable getBackground();

    Drawable[] getIcons();

    /**
     * Tab toggle Icon
     *
     * @param defaultIcon
     * @param selectedIcon
     */
    Tab setIcon(@DrawableRes int defaultIcon, @DrawableRes int selectedIcon);

    /**
     * Tab toggle Icon
     *
     * @param defaultIcon
     * @param selectedIcon
     */
    Tab setIcon(Drawable defaultIcon, Drawable selectedIcon);

    void setTabPadding(int left, int top, int right, int bottom);

    void transition(int textTransitionMode, float positionOffset);

    View getView();

    void setSelected(boolean b);

    /**
     * Tab's Title Color Switching Mode
     *
     * @param textTransitionMode
     */
    void setTextTransitionMode(int textTransitionMode);

    /**
     * Setting Tab Title Selection Time-Varying Coarseness
     *
     * @param b
     */
    Tab setBold(boolean b);

    Tab setTextSize(float size);

    /**
     * Setting the font color of Badge
     *
     * @param color
     */
    Tab setBadgeTextColor(@ColorInt int color);

    /**
     * Set the font size of Badge
     *
     * @param size
     */
    Tab setBadgeTextSize(float size);

    /**
     * Setting the background color of Badge
     *
     * @param color
     */
    Tab setBadgeColor(@ColorInt int color);

    /**
     * Set the border and color of the Badge
     *
     * @param width
     * @param color
     */
    Tab setBadgeStroke(int width, @ColorInt int color);

    void showBadgeMsg(String msg, boolean showDot);

    void hideBadgeMsg();
}

File: tablayout3x/src/main/java/com/liang/tablayout3x/Tab.kt
package com.liang.tablayout3x

import android.content.res.ColorStateList
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.support.annotation.ColorInt
import android.support.annotation.DrawableRes
import android.view.View
import com.liang.widget.TabLayout


interface Tab {
    val contentWidth: Int

    var title: CharSequence?
    var tabTitleColors: ColorStateList?

    var defaultIcon: Drawable?
    var normalIcon: Drawable?
    var selectedIcon: Drawable?
    var tabIconTint: ColorStateList?

    val view: View
    var position: Int
    var tabLayout: TabLayout?

    var tabTitleSize: Float

    var tabRippleColor: Int
    var tabIconTintMode: PorterDuff.Mode?
    var tabUnboundedRipple: Boolean
    var isTabTitleBold: Boolean
    var inlineLabel: Boolean
    var tabBackgroundResId: Int
    var tabDecoration: ItemDecoration?


    fun setTitle(title: String): Tab

    fun setTitleColor(@ColorInt normalColor: Int, @ColorInt selectedColor: Int): Tab

    fun setTitleColor(textColor: ColorStateList?): Tab

    fun setTitleSize(sizePx: Float): Tab

    fun setTitleBold(isBold: Boolean): Tab

    fun setIcon(@DrawableRes icon: Int): Tab

    fun setIcon(icon: Drawable?): Tab

    fun setIconTint(@ColorInt normalColor: Int, @ColorInt selectedColor: Int): Tab

    fun setIconTint(tabIconTint: ColorStateList): Tab

    fun setIcon(@DrawableRes normalIcon: Int, @DrawableRes selectedIcon: Int): Tab

    fun setIcon(normalIcon: Drawable?, selectedIcon: Drawable?): Tab

    fun setRippleColor(tabRippleColor: Int): Tab

    fun setBackground(@DrawableRes resId: Int): Tab

    fun setUnboundedRipple(unboundedRipple: Boolean): Tab

    fun setDecoration(itemDecoration: ItemDecoration): Tab

    fun setTabPadding(
        tabPaddingStart: Int,
        tabPaddingTop: Int,
        tabPaddingEnd: Int,
        tabPaddingBottom: Int
    ): Tab

    fun select()

    fun isSelected(): Boolean

    fun reset()

    fun showBadge(msg: String?)

    fun hideBadge()

    fun setBadgeTextColor(color: Int)

    fun setBadgeTextSize(sp: Float)

    fun setBadgeBackgroundColor(color: Int)

    fun drawBackground(canvas: Canvas)

    fun setBadgeStroke(width: Int, color: Int)

    fun updateColor(offset: Float)

    fun updateScale(scale: Float)
}


interface Badge {
    fun show(msg: String?)

    fun hide()

    fun setBadgeTextSize(sp: Float)

    fun setBadgeBackgroundColor(color: Int)

    fun setBadgeStroke(width: Int, color: Int)

    fun setBadgeTextColor(color: Int)
}


abstract class ItemDecoration(width: Int = 0, height: Int = 0) {

    private val outRect: Rect = Rect()

    var width: Int = width
        get() = if (field == 0) outRect.width() else field
    var height: Int = height
        get() = if (field == 0) outRect.height() else field

    protected val paint: Paint = Paint()

    init {
        paint.isAntiAlias = true
        getItemOffsets(outRect)
    }

    open fun onDraw(canvas: Canvas, rect: Rect) {
    }

    open fun getItemOffsets(outRect: Rect) {
    }
}


interface TabAdapter {
    fun getTab(position: Int): Tab
}




